### ISSUE TITLE
Missing a correct interrupt hook

### ISSUE THREAD

**Issue by Eiyeron** [ndless-nspire/Ndless#23]
Hi. I worked on my own game engine and I actually need the interruptions to work for it to allow the engine to make a robust 60 FPS triple buffering (because !!SCIENCE!!).

With Streetwalrus (that part is mostly that work), we came to overtake some part of the code to hook our own interrupt, blocking us from syscalls and file I/O and also, for unknown reasons, it doesn't work sometimes, the interruptions desyncs, the screen flickers and once the program exists (and disables the interruption hook) triggers a reboot. Also, the interruption code is the only thing that limit us from making a non-CX version of the code.

So,

**Comment by Streetwalrus** [comment]
I don't think the non CX LCD controller supports vertical sync signals, and that isn't the only reason why we weren't supporting it, it's mostly because we don't have hardware to test.
The desync issue might need investigation, I don't know what causes it but I don't think it's because we kill the OS's interrupt handler, that doesn't even run in Ndless programs usually.
The desired API for the handler is nothing too fancy, something like `register(source, handler)` where `source` is the interrupt device ID and `handler` is a function pointer. Alternatively only register one global handler and pass the source to it as an argument. Then of course a cleanup call to uninstall all custom handlers.

**Comment by Vogtinator** [comment]
The method with using cpsr for enabling and disabling IRQs doesn't really work, the SWI handler resets the flags... Not doing that in the handler would be broken as well, for other reasons. Also keep in mind that while the handler is registered, system calls may go horribly wrong.

That shouldn't cause those desyncs though. What happens when run on firebird?

I actually never got a register(source, handler) function to work reliably, but contributions are always welcome. The only time I got IRQs working reliably is pwmtest (https://www.omnimaga.org/ti-nspire-projects/nspire-audio-player!/?action=dlattach;attach=14756).

**Comment by Streetwalrus** [comment]
Yeah the way we do it currently is to uninstall our interrupt handler before every syscall and then reinstall it afterwards.
I haven't tried it on firebird, but with nspire_emu (over a year ago), the LCD controller didn't seem to properly emulated vertical sync signals. It could also be a timer issue. It's hard to reproduce, I still have no idea what triggers it.

**Comment by Eiyeron** [comment]
Last time I tried on firebird it just went directly blinking (so not following the VBlank). Tried with my own Nspire CX CAS's flash and OS.

**Comment by Vogtinator** [comment]
firebird uses the VSYNC of the LCD to record GIFs, it's not synchronized to the GUI though.

**Comment by Eiyeron** [comment]
Indeed, the outputed gif doesn't suffer from the lack of vsync on start. The result GIF is choppier though. I'm trying to run the emulator for a while to wait for the interrupt to provoke this problem. From what I could understand from executing the programs, it looks like the handler messes up after a while, like 2minutes to longer.
